presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/SymbolMapper.java

```diff
@@ -68,74 +70,108 @@ public Expression rewriteSymbolReference(SymbolReference node, Void context, Exp
+        }
 
-    public TopNNode map(TopNNode node, PlanNode source, PlanNodeId newNodeId)
-    {
-        ImmutableList.Builder<Symbol> symbols = ImmutableList.builder();
-        ImmutableMap.Builder<Symbol, SortOrder> orderings = ImmutableMap.builder();
-        Set<Symbol> seenCanonicals = new HashSet<>(node.getOrderBy().size());
-        for (Symbol symbol : node.getOrderBy()) {
-            Symbol canonical = map(symbol);
-            if (seenCanonicals.add(canonical)) {
-                seenCanonicals.add(canonical);
-                symbols.add(canonical);
-                orderings.put(canonical, node.getOrderings().get(symbol));
+        @Override
+        public PlanNode visitTopN(TopNNode node, RewriteContext<Void> context)
+        {
+            PlanNode source = getOnlyElement(recursionStrategy.process(node, context));
+
+            ImmutableList.Builder<Symbol> symbols = ImmutableList.builder();
+            ImmutableMap.Builder<Symbol, SortOrder> orderings = ImmutableMap.builder();
+            Set<Symbol> seenCanonicals = new HashSet<>(node.getOrderBy().size());
+            for (Symbol symbol : node.getOrderBy()) {
+                Symbol canonical = map(symbol);
+                if (seenCanonicals.add(canonical)) {
+                    seenCanonicals.add(canonical);
+                    symbols.add(canonical);
+                    orderings.put(canonical, node.getOrderings().get(symbol));
+                }
             }
+
+            return new TopNNode(
+                    planNodeIdAllocationStrategy.getId(node),
+                    source,
+                    node.getCount(),
+                    symbols.build(),
+                    orderings.build(),
+                    node.getStep());
         }
 
-        return new TopNNode(
-                newNodeId,
-                source,
-                node.getCount(),
-                symbols.build(),
-                orderings.build(),
-                node.getStep());
-    }
+        @Override
+        public PlanNode visitAggregation(AggregationNode node, RewriteContext<Void> context)
+        {
+            PlanNode source = getOnlyElement(recursionStrategy.process(node, context));
 
-    private List<Symbol> mapAndDistinct(List<Symbol> outputs)
-    {
-        Set<Symbol> added = new HashSet<>();
-        ImmutableList.Builder<Symbol> builder = ImmutableList.builder();
-        for (Symbol symbol : outputs) {
-            Symbol canonical = map(symbol);
-            if (added.add(canonical)) {
-                builder.add(canonical);
+            ImmutableMap.Builder<Symbol, Aggregation> aggregations = ImmutableMap.builder();
+            for (Map.Entry<Symbol, Aggregation> entry : node.getAggregations().entrySet()) {
+                Symbol symbol = entry.getKey();
+                Aggregation aggregation = entry.getValue();
+
+                aggregations.put(map(symbol), new Aggregation(
```

Comment:
currently planner should not produce such plan, as symbol allocator will encounter a conflict and so it will look like:
```
project ([a1: b, c: 3 + b],
  aggregation a = count, by b
    values (b)
```